### ISSUE TITLE
Different INVITE From tag after 407 Proxy authentication challenge

### ISSUE THREAD

**Issue by spointer** [davehorton/drachtio-server#106]
Hello, Dave.

When UAC(drachtio-srf) creates and sends a new INVITE after 407 proxy authentication challenge to a remote server, the tag of From header is different from the tag of the original INVITE.

So, the remote server sends 407 Proxy authentication again and fails INVITE as a result.
Can you give me some advice?
Thanks in advance.

Here is the sip log.

2019-11-25 02:38:11.931219 send 1170 bytes to udp/[121.255.185.19]:3551 at 02:38:11.931012:
INVITE sip:3200@121.255.185.19:3551 SIP/2.0
Via: SIP/2.0/UDP 172.25.0.2:3601;rport;branch=z9hG4bKS2v4UQv8e2X0S
Max-Forwards: 70
From: <sip:2001@172.25.0.2:3601>;tag=tmKZvt23B0KZD
To: <sip:3200@121.255.185.19:3551>
Call-ID: 76041f11-a9a2-4950-809f-cc4f9fe9edb8@c-21a783234a415cc1
CSeq: 12770953 INVITE
Contact: <sip:2001@172.25.0.2:3601>
Content-Type: application/sdp
Content-Length: 278

v=0
o=- 2691461379 4 IN IP4 121.182.10.18
s=c
c=IN IP4 121.182.10.18
t=0 0
m=audio 10366 RTP/AVP 0 8 9 101
a=rtpmap:0 PCMU/8000
a=rtpmap:8 PCMA/8000
a=rtpmap:9 G722/8000
a=rtpmap:101 telephone-event/8000
a=fmtp:101 0-16
a=ptime:20
a=maxptime:150
a=sendrecv

2019-11-25 02:38:11.943314 recv 322 bytes from udp/[121.255.185.19]:3551 at 02:38:11.943243:
SIP/2.0 100 Trying
Via: SIP/2.0/UDP 172.25.0.2:3601;received=121.182.10.18;branch=z9hG4bKS2v4UQv8e2X0S;rport=1024
From: <sip:2001@172.25.0.2:3601>;tag=tmKZvt23B0KZD
To: <sip:3200@121.255.185.19:3551>
Call-ID: 76041f11-a9a2-4950-809f-cc4f9fe9edb8@c-21a783234a415cc1
CSeq: 12770953 INVITE
Content-Length: 0

2019-11-25 02:38:11.944525 recv 514 bytes from udp/[121.255.185.19]:3551 at 02:38:11.944436:
SIP/2.0 407 Proxy Authentication Required
Via: SIP/2.0/UDP 172.25.0.2:3601;received=121.182.10.18;branch=z9hG4bKS2v4UQv8e2X0S;rport=1024
From: <sip:2001@172.25.0.2:3601>;tag=tmKZvt23B0KZD
To: <sip:3200@121.255.185.19:3551>;tag=42582aeb5de24811afc190ef84c38dab
Call-ID: 76041f11-a9a2-4950-809f-cc4f9fe9edb8@c-21a783234a415cc1
CSeq: 12770953 INVITE
Proxy-Authenticate: Digest nonce="57e22c9c6e214958a22ee357298d6ae8:d102376b693d4d51a0f044b50331128e",algorithm=MD5,realm="NetDial"
Content-Length: 0

2019-11-25 02:38:11.944954 send 369 bytes to udp/[121.255.185.19]:3551 at 02:38:11.944828:
ACK sip:3200@121.255.185.19:3551 SIP/2.0
Via: SIP/2.0/UDP 172.25.0.2:3601;rport;branch=z9hG4bKS2v4UQv8e2X0S
Max-Forwards: 70
From: <sip:2001@172.25.0.2:3601>;tag=tmKZvt23B0KZD
To: <sip:3200@121.255.185.19:3551>;tag=42582aeb5de24811afc190ef84c38dab
Call-ID: 76041f11-a9a2-4950-809f-cc4f9fe9edb8@c-21a783234a415cc1
CSeq: 12770953 ACK
Content-Length: 0

2019-11-25 02:38:11.950645 send 1403 bytes to udp/[121.255.185.19]:3551 at 02:38:11.950411:
INVITE sip:3200@121.255.185.19:3551 SIP/2.0
Via: SIP/2.0/UDP 172.25.0.2:3601;rport;branch=z9hG4bKtBpXXjDccBmKN
Max-Forwards: 70
From: <sip:2001@172.25.0.2:3601>;tag=UXcryNK7889HS
To: <sip:3200@121.255.185.19:3551>
Call-ID: 76041f11-a9a2-4950-809f-cc4f9fe9edb8@c-21a783234a415cc1
CSeq: 12770954 INVITE
Contact: <sip:2001@172.25.0.2:3601>
Proxy-Authorization: Digest username="2001", realm="NetDial", nonce="57e22c9c6e214958a22ee357298d6ae8:d102376b693d4d51a0f044b50331128e", uri="sip:3200@121.255.185.19:3551", response="6afcc87765a176d48ba2d6b11ac456c9", algorithm=MD5
Content-Type: application/sdp
Content-Length: 278

v=0
o=- 2691461379 4 IN IP4 121.182.10.18
s=c
c=IN IP4 121.182.10.18
t=0 0
m=audio 10366 RTP/AVP 0 8 9 101
a=rtpmap:0 PCMU/8000
a=rtpmap:8 PCMA/8000
a=rtpmap:9 G722/8000
a=rtpmap:101 telephone-event/8000
a=fmtp:101 0-16
a=ptime:20
a=maxptime:150
a=sendrecv

2019-11-25 02:38:11.963030 recv 322 bytes from udp/[121.255.185.19]:3551 at 02:38:11.962961:
SIP/2.0 100 Trying
Via: SIP/2.0/UDP 172.25.0.2:3601;received=121.182.10.18;branch=z9hG4bKtBpXXjDccBmKN;rport=1024
From: <sip:2001@172.25.0.2:3601>;tag=UXcryNK7889HS
To: <sip:3200@121.255.185.19:3551>
Call-ID: 76041f11-a9a2-4950-809f-cc4f9fe9edb8@c-21a783234a415cc1
CSeq: 12770954 INVITE
Content-Length: 0

2019-11-25 02:38:11.964378 recv 514 bytes from udp/[121.255.185.19]:3551 at 02:38:11.964298:
SIP/2.0 407 Proxy Authentication Required
Via: SIP/2.0/UDP 172.25.0.2:3601;received=121.182.10.18;branch=z9hG4bKtBpXXjDccBmKN;rport=1024
From: <sip:2001@172.25.0.2:3601>;tag=UXcryNK7889HS
To: <sip:3200@121.255.185.19:3551>;tag=7c7a322113044568938b3959c306af63
Call-ID: 76041f11-a9a2-4950-809f-cc4f9fe9edb8@c-21a783234a415cc1
CSeq: 12770954 INVITE
Proxy-Authenticate: Digest nonce="68421bf800f34ad99cc6ff5913f18ce2:a239a59fa582449c8f97a8bbc804a9b2",algorithm=MD5,realm="NetDial"
Content-Length: 0

**Comment by davehorton** [comment]
can you confirm what version you are using ?

**Comment by spointer** [comment]
drachtio-server version is v0.8.3-4-g8fe175fb0 and drachtio-srf is 4.4.7

**Comment by davehorton** [comment]
can you try with the latest drachtio-srf (no need to upgrade the server)?

**Comment by spointer** [comment]
I upgraded drachtio-srf to the latest version(4.4.22), but the result is same.
I sent a log file to your email with loglevel to debug and sofia-loglevel to 9

Thank you.

**Comment by davehorton** [comment]
This is strange because I have a test case under drachtio-srf and it passes.  Having trouble recreating your issue.  Can you update to latest drachtio server and try with that?  If you are able to create there, I may need you to provide a simple app

**Comment by davehorton** [comment]
specifically, this line in digest-client.js is supposed to maintain that tag:
```
      // preserve tag on From header as well
      headers['From'] = this.req.get('from') ;
```

**Comment by spointer** [comment]
Hi Dave,

I updated drachtio-server to the latest version(v0.8.3-10-g71359a14a) and tested again, but had no luck.

I found some strange behavior of drachtio-srf and drachtio-server, so sent it to your email.

To sum up from logs, it is as follows.

1. drachtio-srf sends From header ==> from: sip:2001@localhost
2. drahctio-server sends actual From header ==> From: <sip:2001@172.25.0.2:3601>;tag=8p64vKQ149p2m
3. drachtio-srf received 407
4. drachtio-srf sends two From header ==> from: sip:2001@localhost and From: <sip:2001@172.25.0.2:3601>;tag=8p64vKQ149p2m
5. drachtio-server sends actual From header with different tag ==> From: <sip:2001@172.25.0.2:3601>;tag=9ZZXye841jDNg

I think 1, 2, 3 are correct but 4, 5 is somewhat strange.

Thank you.
